### DISCUSSION TITLE
Q: Recommended database parameters for write intensive setups

### DISCUSSION THREAD

**Discussion by bitmorph** [erthink/libmdbx#344] [Q&A] [answered]
First off, thanks a lot for your great work and many contributions to the open-source ecosystem.

I'd like to use `libmdbx` as a simple time series storage (where the keys are timestamps and values are some serialized objects). A quick and dirty prototype--I used [mdbx-go](https://github.com/torquem-ch/mdbx-go) if it matters--with default parameters (`MDBX_ENV_DEFAULTS`) gave me tons of `MDBX_MAP_FULL` errors when writing data intensively. Then I played with flags like `MDBX_WRITEMAP`, `MDBX_SAFE_NOSYNC` etc. but saw the same errors, as well. The fact that I just can't write straightforward working code led me here eventually.

I guess such tough-write cases need a kind of tuning of the database with environment flags and `mdbx_env_set_geometry` but can't figure out which exactly I should pick and what trade-offs with reading data may arise.

Some rapid googling was in vain. Could you please give me a reference or a bunch of advice to make informed choices about write-only/read-only and write-read parameters for `libmdbx` gaining reasonable performance?

**Comment by erthink** [comment, marked as answer]
> First off, thanks a lot for your great work and many contributions to the open-source ecosystem.

Thank you for support.

> ... gave me tons of MDBX_MAP_FULL errors ...

Basically you should use `mdbx_env_set_geometry()` to setup maximal DB size.
Otherwise a historical default will be used which is a few megabytes only.

> ... MDBX_SAFE_NOSYNC ... Some rapid googling was in vain.

Please read the [description of MDBX_SAFE_NOSYNC](https://libmdbx.dqdkfa.ru/group__c__opening.html#gga9138119a904355d245777c4119534061a586d75ff2fd68aed647a60ea7984ed85) and, if necessary, ask a question about what exactly you do not understand.
At the very least, this will help clarify the relevant descriptions.

> Could you please give me a reference or a bunch of advice to make informed choices about write-only/read-only and write-read parameters for libmdbx gaining reasonable performance?

First, please read [the introduction](https://libmdbx.dqdkfa.ru/intro.html) carefully, at least once in full.
I hope this will help you clarify many points.

Briefly:
- MDBX is based on [B+Tree](https://en.wikipedia.org/wiki/B%2B_tree), but does not use [WAL](https://en.wikipedia.org/wiki/Write-ahead_logging).
- Therefore, every update in MDBX requires OLog(database size) of IOPs (random page writes).
- So MDBX was not designed to provide high write performance simultaneously with the durability, at least on a HDDs or slow SSDs.

MDBX is super-good for read-only and mostly-read cases, but for write performance you should use relevant hardware, or agreed significant risks of reliability in case of system crashes (power outages or fatal failure of the OS kernel, but not an application crash):
- `MDBX_SYNC_DURABLE` with a high-performance NVMe based on persistent memory (i.e [NVDIMM](https://en.wikipedia.org/wiki/NVDIMM), etc).
- `MDBX_SYNC_DURABLE | MDBX_LIFORECLAIM` with a good-performance disk with battery/flash-backed write-cache (i.e. BBWC or FBWC).
- `MDBX_SYNC_DURABLE` with a high-end NVMe or SSD.
- performance/durability tradeoff: `MDBX_SAFE_NOSYNC` with the background thread calling `mdbx_env_sync_ex()` in a loop.
In the worst case scenario, you will lose the last few transactions. On the other hand, this may be a good solution if a dedicated disk will be used for the DB.
- no durability at all: `MDBX_UTTERLY_NOSYNC`, but in case of power outages or fatal failure of the OS kernel (except an application crash) you will lose whole DB.

**Comment by bitmorph** [comment]
@erthink, many thanks for the prompt and detailed reply. I'm going to dive into the documentation armed with your helpful explanations. I'll close this issue to avoid polluting the issues board with general questions and open new ones about specific problems.

**Comment by erthink** [comment]
https://www.microsemi.com/document-portal/doc_view/135965-flashtec-a-nvram-drive-family
ebay (this is a limited offer): [PCI-E NVRAM Drive](https://www.ebay.com/itm/254346621402) + [Battery](https://www.ebay.com/itm/264447756501) = ~$200 for ~1.000.000 TPS with durability (by libmdbx on Linux).
